cli-plugins/manager/hooks.go

```diff
@@ -0,0 +1,180 @@
+}
+
+// TemplateReplaceArg takes an index i and returns a hook
+// template string that the CLI will replace the template with
+// the ith argument, after processing the passed flags.
+//
+// Example:
+//
+// "run this image with `docker run " + TemplateReplaceArg(0) + "`"
+//
+// when being executed after the command:
+// `docker pull alpine`
+// will result in the message:
+// "Run this image with `docker run alpine`"
+func TemplateReplaceArg(i int) string {
+	return fmt.Sprintf(hookTemplateArg, strconv.Itoa(i))
+}
+
+// RunPluginHooks calls the hook subcommand for all present
+// CLI plugins that declare support for hooks in their metadata
+// and parses/prints their responses
+func RunPluginHooks(dockerCli command.Cli, rootCmd, subCommand *cobra.Command) error {
+	plugins, err := ListPlugins(dockerCli, rootCmd)
+	if err != nil {
+		return err
+	}
+	subCmdName := subCommand.Name()
+
+	hookMessages := make([]string, 0, len(plugins))
+	for _, p := range plugins {
+		p := p
+		if !p.supportsHookCommand(subCmdName) {
+			continue
+		}
+
+		hookReturn, err := p.RunHook(subCmdName)
+		// Skip misbehaving plugins, but don't halt execution
+		if err != nil {
+			continue
+		}
+		hookMessages = append(hookMessages, parseHooks(hookReturn.Template, subCommand))
+	}
+
+	printHookMessages(dockerCli.Err(), hookMessages)
+	return nil
+}
+
+const (
+	hookTemplateCommandName = "{docker-cli-plugin-hook-cmd-name}"
+	hookTemplateFlagValue   = "{docker-cli-plugin-hook-flag-%s}"
+	hookTemplateArg         = "{docker-cli-plugin-hook-arg-%s}"
+)
+
+func parseHooks(template string, cmd *cobra.Command) string {
+	ret := template
+	for {
+		parsed, foundMatch := parseFlagValue(ret, cmd)
+		ret = parsed
+		if !foundMatch {
+			break
+		}
+	}
+	for {
+		parsed, foundMatch := parseArg(ret, cmd)
+		ret = parsed
+		if !foundMatch {
+			break
+		}
+	}
+	return strings.ReplaceAll(ret, hookTemplateCommandName, cmd.Name())
+}
+
+var (
+	flagRegex = regexp.MustCompile(fmt.Sprintf(hookTemplateFlagValue, `([\w-]+)`))
+	argRegex  = regexp.MustCompile(fmt.Sprintf(hookTemplateArg, `([\w-]+)`))
+)
+
+func parseFlagValue(template string, cmd *cobra.Command) (string, bool) {
+	matches := flagRegex.FindStringSubmatch(template)
+	if len(matches) == 0 {
+		return template, false
+	}
+	for _, m := range matches[1:] {
+		flag := cmd.Flag(m)
+		if flag == nil {
+			continue
+		}
+		value := flag.Value.String()
+		template = strings.ReplaceAll(template, TemplateReplaceFlagValue(m), value)
+	}
+	return template, true
+}
+
+func parseArg(template string, cmd *cobra.Command) (string, bool) {
+	matches := argRegex.FindStringSubmatch(template)
+	if len(matches) == 0 {
+		return template, false
+	}
+	for _, m := range matches[1:] {
+		i, err := strconv.Atoi(m)
+		if err != nil {
+			continue
+		}
+		if i >= len(cmd.Flags().Args()) {
+			return template, false
+		}
+		template = strings.ReplaceAll(template, TemplateReplaceArg(i), cmd.Flags().Arg(i))
+	}
+	return template, true
```

Comment:
I'm not so sure about this, returning true here wouldn't necessarily mean that the matches we found were valid and actually had an impact on the template string.  

`matches` could have only 1 match, leading to the for being ignored, or `strconv` could've failed exiting the for loop (if no more matches) ending up with us returning the template string and `true`.

Reply:
Did a large refactor here, PTAL and let me know if you like the new implementation better :)